File: lib/agents/state.ts
// CurrentState — shared "nervous system" of the agent network.
// Pure functions only (server-safe). No localStorage access here.
import type { Event } from "@/lib/types";

export type Mood = "anxious" | "energized" | "avoidant" | "focused" | "burnt-out" | "neutral";
export type ToneMode = "gentle" | "energetic" | "stoic";
export type SessionPhase = "idle" | "working" | "break" | "planning";

export type AgentTrailEntry = { agent: string; ts: number; summary: string };

export type CurrentState = {
  mood: Mood;
  energy: number; // 0-1
  toneMode: ToneMode;
  activeTaskId?: string;
  sessionStart: number;
  phase: SessionPhase;
  urgencyTop3: string[]; // task IDs by TMT score
  agentTrail: AgentTrailEntry[];
  lastUpdated: number;
};

// Extended context that all internal agents receive (superset of AgentContext)
import type { Message, Profile, BeliefState, Task } from "@/lib/types";

export type HelmContext = {
  input: string;
  messages: Message[];
  profile?: Profile;
  beliefs?: BeliefState;
  tasks?: Task[];
  events?: Event[];
  state: CurrentState;
};

export const DEFAULT_STATE: CurrentState = {
  mood: "neutral",
  energy: 0.7,
  toneMode: "gentle",
  sessionStart: Date.now(),
  phase: "idle",
  urgencyTop3: [],
  agentTrail: [],
  lastUpdated: Date.now(),
};

export function createState(patch?: Partial<CurrentState>): CurrentState {
  return { ...DEFAULT_STATE, ...patch, lastUpdated: Date.now() };
}

export function patchState(state: CurrentState, patch: Partial<CurrentState>): CurrentState {
  return { ...state, ...patch, lastUpdated: Date.now() };
}

export function addToTrail(state: CurrentState, entry: AgentTrailEntry): CurrentState {
  const trail = [...state.agentTrail.slice(-19), entry];
  return { ...state, agentTrail: trail, lastUpdated: Date.now() };
}

export function moodToTone(mood: Mood): ToneMode {
  if (mood === "anxious" || mood === "burnt-out") return "gentle";
  if (mood === "avoidant" || mood === "focused") return "stoic";
  return "energetic";
}

// Fast keyword-based mood inference (no LLM cost)
export function inferMoodFromText(text: string): Mood {
  const t = text.toLowerCase();
  if (/tired|exhaust|drain|can't|cannot|overwhelm|stuck|lost|burnout|burnt/.test(t)) return "burnt-out";
  if (/anxious|worried|scared|nervous|panic|stress|pressure|deadline/.test(t)) return "anxious";
  if (/avoid|procrastinat|delay|later|don't feel|not feel|can't start|can't begin/.test(t)) return "avoidant";
  if (/ready|let's go|pump|excit|fire|grind|crush|let me|going to/.test(t)) return "energized";
  if (/focus|working|in flow|deep work|concentrat|on it/.test(t)) return "focused";
  return "neutral";
}

export function inferEnergyFromText(text: string): number {
  const t = text.toLowerCase();
  if (/tired|exhaust|drain|burnt|can't|overwhelm|stuck/.test(t)) return 0.2;
  if (/energiz|ready|pump|excit|flow|crush/.test(t)) return 0.9;
  if (/okay|fine|alright|so-so|meh/.test(t)) return 0.5;
  return 0.65;
}

File: backend/avnik/state.py
"""CurrentState helpers — the agent network's nervous system.

Pure functions, faithful port of lib/agents/state.ts.
"""
import re
from .models import CurrentState, AgentTrailEntry, now_ms


def infer_mood_from_text(text: str) -> str:
    t = text.lower()
    if re.search(r"tired|exhaust|drain|can't|cannot|overwhelm|stuck|lost|burnout|burnt", t):
        return "burnt-out"
    if re.search(r"anxious|worried|scared|nervous|panic|stress|pressure|deadline", t):
        return "anxious"
    if re.search(r"avoid|procrastinat|delay|later|don't feel|not feel|can't start|can't begin", t):
        return "avoidant"
    if re.search(r"ready|let's go|pump|excit|fire|grind|crush|let me|going to", t):
        return "energized"
    if re.search(r"focus|working|in flow|deep work|concentrat|on it", t):
        return "focused"
    return "neutral"


def infer_energy_from_text(text: str) -> float:
    t = text.lower()
    if re.search(r"tired|exhaust|drain|burnt|can't|overwhelm|stuck", t):
        return 0.2
    if re.search(r"energiz|ready|pump|excit|flow|crush", t):
        return 0.9
    if re.search(r"okay|fine|alright|so-so|meh", t):
        return 0.5
    return 0.65


def mood_to_tone(mood: str) -> str:
    if mood in ("anxious", "burnt-out"):
        return "gentle"
    if mood in ("avoidant", "focused"):
        return "stoic"
    return "energetic"


def add_to_trail(state: CurrentState, agent: str, summary: str) -> CurrentState:
    trail = state.agentTrail[-19:] + [AgentTrailEntry(agent=agent, ts=now_ms(), summary=summary[:120])]
    state.agentTrail = trail
    state.lastUpdated = now_ms()
    return state


def regulate(state: CurrentState, user_input: str) -> CurrentState:
    """Step 1 of the loop: infer mood/energy/tone from the latest message."""
    state.mood = infer_mood_from_text(user_input)
    state.energy = infer_energy_from_text(user_input)
    state.toneMode = mood_to_tone(state.mood)
    state.lastUpdated = now_ms()
    return state
